tools/flow_grid.py

```diff
@@ -0,0 +1,689 @@
+                dat = data[corner[i]['k']]
+                sur = data[corner[i]['v']]
+                if ((dat - sur) > 0).any():
+                    outmap[corner[i]['k']] = corner[i]['pad'][np.argmax(dat - sur)]
+                else:
+                    outmap[corner[i]['k']] = flat
+
+            # fill edges
+            for x in edge.keys():
+                dat = data[edge[x]['k']]
+                sur = data[select_edge_sur(x)]
+                a = ((dat - sur) > 0).any(axis=0)
+                b = edge[x]['pad'][np.argmax((dat - sur), axis=0)]
+                c = flat
+                outmap[edge[x]['k']] = np.where(a,b,c)
+    
+        # If direction numbering isn't default, convert values of output array.
+        if dirmap != (1,2,3,4,5,6,7,8):
+            dir_d = dict(zip((1,2,3,4,5,6,7,8), dirmap))
+            outmap = pd.DataFrame(outmap).apply(lambda x: x.map(dir_d), axis=1).values
+
+        np.place(outmap, dem_mask, nodata)
+        np.place(data, dem_mask, dem_nodata)        
+        
+        if inplace == True:
+            self.dir = outmap
+        else:
+            return outmap
+
+    def catchment(self, x, y, pour_value=None, dirmap=(1,2,3,4,5,6,7,8),
+                  nodata=0, xytype='index', recursionlimit=15000, inplace=True):
+        """
+        Delineates a watershed from a given pour point (x, y).
+        Returns a grid 
+
+        Parameters
+        ----------
+        x : int or float
+            x coordinate of pour point
+        y : int or float
+            y coordinate of pour point
+        pour_value : int or None
+                     If not None, value to represent pour point in catchment
+                     grid (required by some programs).
+        dirmap : list or tuple
+                 List of integer values representing the following
+                 cardinal directions (in order):
+                 [N, NE, E, SE, S, SW, W, NW] 
+        xytype : 'index' or 'label'
+                 How to interpret parameters 'x' and 'y'.
+                     'index' : x and y represent the column and row 
+                               indices of the pour point.
+                     'label' : x and y represent geographic coordinates
+                               (will be passed to self.nearest_cell).
+        recursionlimit : int
+                         Recursion limit--may need to be raised if
+                         recursion limit is reached.
+        inplace : bool
+                  If True, catchment will be written to attribute 'catch'.
+        """
+
+        # if xytype is 'label', delineate catchment based on cell nearest
+        # to given geographic coordinate
+        if xytype == 'label':
+            x, y = self.nearest_cell(x, y)
+
+        # set recursion limit (needed for large datasets)
+        sys.setrecursionlimit(recursionlimit)
+
+        # initialize array to collect catchment cells
+        self.collect = np.array([], dtype=int)
```

Comment:
I think this is the first approach I tried, and it was giving me problems. I can't remember what the specific problem was because I wrote this part about a month ago. I'll have to go back and take a look at it.